### ISSUE TITLE
The question about reproduce RC-SQD results

### ISSUE THREAD

**Issue by kugwzk** [princeton-nlp/DensePhrases#11]
Hi~ Thanks a lot for your open source work.
When I run your code for SQuAD dataset in one passage training, I got 77.3 EM and 85.7 F1. I ran code in this script`    - python train_rc.py --model_type bert
      --pretrained_name_or_path SpanBERT/spanbert-base-cased
      --data_dir densephrases/densephrases-data/single-qa
      --cache_dir densephrases/cache
      --train_file squad/train-v1.1_qg_ents_t5large_3500_filtered.json
      --predict_file squad/dev-v1.1.json
      --do_train --do_eval
      --per_gpu_train_batch_size 24
      --learning_rate 3e-5
      --num_train_epochs 3.0
      --max_seq_length 384
      --seed 42
      --fp16
      --fp16_opt_level O1
      --lambda_kl 4.0
      --lambda_neg 2.0
      --lambda_flt 1.0
      --filter_threshold -2.0
      --append_title
      --evaluate_during_training
      --overwrite_output_dir
      --teacher_dir densephrases/outputs/spanbert-base-cased-squad`
I also train this model for another 2 epochs like your makefile using pre-batch negative and train-v1.1.json (the real squad data), but the results is still below the paper results.
(1) Does I should use different hyperparameters? I found your paper use different parameters with your script, such as batch size (84 vs 24) or lambda weight, etc.
(2) In the paper, the results are the average of random seeds?
(3) Do you use the whole nq and squad datasets to train the model?

**Comment by jhyuklee** [comment]
Hi, thanks for posting the issue.
I think the main cause for slight decrease of the performance is 1. using fp16 and 2. using filtering options (--lambda_flt 1.0 and --filter_threshold -2.0) which were not used for the single-passage evaluation.

For your other questions:
(1) Yes, the batch size has been decreased in the open-source code to match the smaller-RAM GPUs (e.g., 24GB).
(2) They are maximum scores, but didn't change much with different seeds.
(3) For singe-passage performances, we train on each dataset separately and evaluate on them.

Let me know if you have more questions.

**Comment by kugwzk** [comment]
Do you mean lambda_flt=0.0 to reproduce the results?

**Comment by jhyuklee** [comment]
Yes, you can set it like that + remove filter_threshold parameter (use the default value). Also note that the current makefile setup is mostly tuned for semi-open-domain and open-domain settings where the number of passages to index are much larger. If you only care about single-passage performance, in-batch negatives might also be unnecessary for SQuAD (lambda_neg = 0).

**Comment by kugwzk** [comment]
Thanks a lot. I have run a job follow your advice, I will give some feedback later.

**Comment by kugwzk** [comment]
Hi, Lee~ I follow your advice and run with script: 
`stage1: python train_rc.py --model_type bert --pretrained_name_or_path SpanBERT/spanbert-base-cased --data_dir /densephrases-data/single-qa --cache_dir /densephrases/cache --train_file $TRAIN_QG_DATA --predict_file squad/dev-v1.1.json --do_train --do_eval --per_gpu_train_batch_size 24 --learning_rate 3e-5 --num_train_epochs 2 --max_seq_length 384 --seed 42 --lambda_kl 4.0 --lambda_neg 2.0 --lambda_flt 0.0 --append_title --evaluate_during_training --overwrite_output_dir --teacher_dir densephrases/outputs/spanbert-base-cased-squad --output_dir $STAGE1_OUTPUT_PATH`
`stage2: python train_rc.py --model_type bert --pretrained_name_or_path SpanBERT/spanbert-base-cased --data_dir densephrases-data/single-qa --cache_dir /densephrases/cache --train_file $TRAIN_DATA --predict_file squad/dev-v1.1.json --do_train --do_eval --per_gpu_train_batch_size 24 --learning_rate 3e-5 --num_train_epochs 2 --max_seq_length 384 --seed 42 --lambda_kl 4.0 --lambda_neg 2.0 --append_title --evaluate_during_training --overwrite_output_dir --teacher_dir densephrases/outputs/spanbert-base-cased-squad --output_dir $STAGE2_OUTPUT_PATH --pbn_size 2 --pbn_tolerance 0 --load_dir $STAGE1_OUTPUT_PATH`
As for stage1, EM/F1 77.4/85.9 and stage2 are 77.6 and 86.3. I found the EM results are still worse than the paper results. Might because I should use a large batch size like paper or remove negative loss?

**Comment by jhyuklee** [comment]
Thanks for running the experiment! Can you just set lambda_neg as 0.0? Single-passage performance and multiple-passage performance are often in a trade-off relationship. First step would be fine with QG.

**Comment by kugwzk** [comment]
Thanks a lot for your advice. I will closed this issue.
BTW, I run experiments w/ or w/o fp16 get similar results.
